rerun_py/src/dataframe.rs

```diff
@@ -0,0 +1,635 @@
+        let end = re_sdk::Time::from_ns_since_epoch(end);
+
+        let resolved = ResolvedTimeRange::new(start, end);
+
+        let mut query_expression = self.query_expression.clone();
+        query_expression.filtered_index_range = Some(resolved);
+
+        Ok(Self {
+            recording: self.recording.clone(),
+            query_expression,
+        })
+    }
+}
+
+impl PyRecording {
+    fn engine(&self) -> QueryEngine<'_> {
+        QueryEngine {
+            store: &self.store,
+            cache: &self.cache,
+        }
+    }
+
+    /// Convert a `ViewContentsLike` into a `ViewContentsSelector`.
+    ///
+    /// ```python
+    /// ViewContentsLike = Union[str, Dict[str, Union[ComponentLike, Sequence[ComponentLike]]]]
+    /// ```
+    ///
+    /// We cant do this with the norma `FromPyObject` mechanisms because we want access to the
+    /// `QueryEngine` to resolve the entity paths.
+    fn extract_contents_expr(
+        &self,
+        expr: Bound<'_, PyAny>,
+    ) -> PyResult<re_chunk_store::ViewContentsSelector> {
+        let engine = self.engine();
+
+        if let Ok(expr) = expr.extract::<String>() {
+            // `str`
+
+            let path_filter = EntityPathFilter::parse_strict(&expr, &Default::default())
+                .map_err(|err| PyValueError::new_err(err.to_string()))?;
+
+            let contents = engine
+                .iter_entity_paths(&path_filter)
+                .map(|p| (p, None))
+                .collect();
+
+            Ok(contents)
+        } else if let Ok(dict) = expr.downcast::<PyDict>() {
+            // `Union[ComponentLike, Sequence[ComponentLike]]]`
+
+            let mut contents = ViewContentsSelector::default();
+            for (key, value) in dict {
+                let key = key.extract::<String>()?;
+
+                let path_filter = EntityPathFilter::parse_strict(&key, &Default::default())
+                    .map_err(|err| PyValueError::new_err(err.to_string()))?;
+
+                let components: BTreeSet<ComponentName> =
+                    if let Ok(component) = value.extract::<ComponentLike>() {
+                        std::iter::once(component.0).collect()
+                    } else if let Ok(components) = value.extract::<Vec<ComponentLike>>() {
+                        components.into_iter().map(|c| c.0).collect()
+                    } else {
+                        return Err(PyTypeError::new_err(
+                            "ViewContentsLike input must be a string or a list of strings.",
+                        ));
+                    };
+
+                contents.append(
+                    &mut engine
+                        .iter_entity_paths(&path_filter)
+                        .map(|p| (p, Some(components.clone())))
+                        .collect(),
+                );
+            }
+
+            Ok(contents)
+        } else {
+            return Err(PyTypeError::new_err("ViewContentsLike input must be..."));
```

Comment:
Is there some pythonic magic going on here or was the err message never finished? :thinking:

Reply:
No magic -- just forgot to finish while I was figuring out what to write.